File: flutter/lib/src/cache/campaign_cache.dart
import 'package:moengage_flutter/moengage_flutter.dart';

import 'i_campaign_cache.dart';

/// In-memory implementation of [ICampaignCache].
///
/// Backed by a plain [Map]; entries are evicted per-campaign on [remove]
/// (post-dismiss) and globally on [clear] (teardown).
///
/// Swap this with an LRU or persistent cache by implementing [ICampaignCache]
/// and injecting it into [MoEngagePlugin] — no other code changes required.
final class CampaignCache implements ICampaignCache {
  final Map<String, SelfHandledCampaignData> _store = {};

  @override
  void put(String campaignId, SelfHandledCampaignData data) =>
      _store[campaignId] = data;

  @override
  SelfHandledCampaignData? get(String campaignId) => _store[campaignId];

  @override
  void remove(String campaignId) => _store.remove(campaignId);

  @override
  void clear() => _store.clear();

  @override
  int get count => _store.length;

  @override
  List<String> get campaignIds => List.unmodifiable(_store.keys);
}

File: android/digia-moengage/src/main/java/com/digia/moengage/cache/CampaignCache.kt
package com.digia.moengage.cache

import com.moengage.inapp.model.SelfHandledCampaignData

/**
 * In-memory implementation of [ICampaignCache].
 *
 * Backed by a plain [MutableMap]; entries are evicted per-campaign on [remove] (post-dismiss) and
 * globally on [clear] (teardown).
 *
 * Swap this with an LRU or persistent cache by implementing [ICampaignCache] and injecting it into
 * [MoEngagePlugin] — no other code changes required.
 */
class CampaignCache : ICampaignCache {

    private val store = mutableMapOf<String, SelfHandledCampaignData>()

    override fun put(campaignId: String, data: SelfHandledCampaignData) {
        store[campaignId] = data
    }

    override fun get(campaignId: String): SelfHandledCampaignData? = store[campaignId]

    override fun remove(campaignId: String) {
        store.remove(campaignId)
    }

    override fun clear() = store.clear()

    override val count: Int
        get() = store.size

    override val campaignIds: List<String>
        get() = store.keys.toList()
}
